### PR TITLE
ARM_COMPUTE_NO_EXCEPTIONS macro guard

### PR DESCRIPTION
**Pull request by lukeiwanski**
Adds ARM_COMPUTE_NO_EXCEPTIONS macro guard for targets that do not accept exceptions

### CODE REVIEW

**Comment by lukeiwanski** [comment]
@GeorgeARM is this better?

File: utils/Utils.h [hunk 1]
```diff
@@ -24,6 +24,16 @@
 #ifndef __UTILS_UTILS_H__
 #define __UTILS_UTILS_H__
 
+#ifdef ARM_COMPUTE_NO_EXCEPTIONS
```
Comment (AnthonyBarbier) [inline utils/Utils.h]: This file is not used by the library: it's only needed for tests / examples, so you shouldn't need to update it ?

File: include/libnpy/npy.hpp [hunk 2]
```diff
@@ -80,11 +80,12 @@ inline void read_magic(std::istream& istream, unsigned char& v_major, unsigned c
   istream.read(buf, magic_string_length+2);
 
   if(!istream) {
-    throw std::runtime_error("io error: failed reading file");
+    ARM_COMPUTE_THROW("io error: failed reading file");
```
Comment (AnthonyBarbier) [inline include/libnpy/npy.hpp]: This is a third party library, we need to upstream the changes in a cleaner way.

File: src/core/Error.cpp [hunk 3]
```diff
@@ -56,6 +56,7 @@ void arm_compute::error(const char *function, const char *file, const int line,
     va_end(args);
     throw std::runtime_error(err.error_description());
 }
+
```
Comment (AnthonyBarbier) [inline src/core/Error.cpp]: Unneeded change

File: utils/Utils.cpp [hunk 4]
```diff
@@ -70,7 +70,6 @@ int run_example(int argc, const char **argv, example &func)
 {
     std::cout << "\n"
               << argv[0] << "\n\n";
-
     try
```
Comment (AnthonyBarbier) [inline utils/Utils.cpp]: Unneeded change

File: utils/Utils.h (hunk 1 above)
Comment (lukeiwanski) [reply utils/Utils.h]: Ah okay, so the really only changes are needed to npy.hpp?
It looked like this header is utility used by the framework.. obviously I was mistaken.
I will revert changes to this file.

File: include/libnpy/npy.hpp (hunk 2 above)
Comment (lukeiwanski) [reply include/libnpy/npy.hpp]: Would that be acceptable to define `ARM_COMPUTE_THROW` in this file? Or do you have something else in mind?

**Comment by lukeiwanski** [comment]
@GeorgeARM  seems like this https://github.com/ARM-software/ComputeLibrary/pull/325#commitcomment-26593339 will not be needed after all

File: include/libnpy/npy.hpp (hunk 2 above)
Comment (AnthonyBarbier) [reply include/libnpy/npy.hpp]: Are you sure this is needed ? 
I just had a quick look and it looks like this is only included in `utils/Utils.h` and `tests/AssetsLibrary.h` ?
So I think if you build with `examples=0 benchmark_tests=0 validation_tests=0` then you should be fine.

File: utils/Utils.h (hunk 1 above)
Comment (AnthonyBarbier) [reply utils/Utils.h]: We have a different `arm_compute/core/Utils.h` but I don't think `utils/Utils.h` is used outside of the examples and tests

File: include/libnpy/npy.hpp (hunk 2 above)
Comment (lukeiwanski) [reply include/libnpy/npy.hpp]: actually it's not needed!
My problem started with wrong assumptions that I need to include `utils/Utils.h`.
It turns out that the only change needed is that to `arm_compute/core/CL/OpenCL.h`

File: utils/Utils.h (hunk 1 above)
Comment (lukeiwanski) [reply utils/Utils.h]: yes, that essentially was cause of my confusion. I can confirm that `utils/Utils.h` is not needed!
I will revert not needed changes.

**Pull request merged**